### PR TITLE
improve self transfer handling

### PR DESCRIPTION
**Pull request by xlc**
improve self transfer handling for transfer_share_and_rewards

### CODE REVIEW

File: rewards/src/tests.rs [hunk 1]
```diff
@@ -593,3 +593,35 @@ fn transfer_share_and_rewards() {
...
+		assert_noop!(
+			RewardsModule::transfer_share_and_rewards(&ALICE, &DOT_POOL, 200, &ALICE),
+			Error::<Runtime>::CanSplitOnlyLessThanShare
+		);
+
+		let pool_info = RewardsModule::pool_infos(DOT_POOL);
+		assert_ok!(RewardsModule::transfer_share_and_rewards(&ALICE, &DOT_POOL, 33, &ALICE));
+
+		let new_pool_info = RewardsModule::pool_infos(DOT_POOL);
+		assert_eq!(pool_info, new_pool_info, "reward transfer does not affect the pool");
```
Comment (zjb0807) [inline rewards/src/tests.rs:616]: Should it be transferred to BOB to verify this case?

File: rewards/src/lib.rs [hunk 2]
```diff
@@ -329,11 +329,15 @@ impl<T: Config> Pallet<T> {
 		move_share: T::Share,
 		other: &T::AccountId,
 	) -> DispatchResult {
-		SharesAndWithdrawnRewards::<T>::mutate(pool, other, |increased_share| {
+		SharesAndWithdrawnRewards::<T>::try_mutate(pool, other, |increased_share| {
 			let (increased_share, increased_rewards) = increased_share;
-			SharesAndWithdrawnRewards::<T>::mutate_exists(pool, who, |share| {
+			SharesAndWithdrawnRewards::<T>::try_mutate_exists(pool, who, |share| {
 				let (share, rewards) = share.as_mut().ok_or(Error::<T>::ShareDoesNotExist)?;
 				ensure!(move_share < *share, Error::<T>::CanSplitOnlyLessThanShare);
```
Comment (ermalkaleci) [inline rewards/src/lib.rs:336]: should we add like ED for shares? It can be a const with min possible amount so transfer or withdrawal won't leave dust

Comment (xlc) [reply rewards/src/lib.rs:336]: yeah in next PR

File: rewards/src/tests.rs (hunk 1 above)
Comment (xlc) [reply rewards/src/tests.rs:616]: this is to test transfer to self, which previously wasn't handled correctly

**Review by zjb0807 — APPROVED** [review]

**Pull request merged**